core/src/test/java/apoc/trigger/TriggerClusterRoutingTest.java

```diff
@@ -0,0 +1,178 @@
+                                "apoc.trigger.enabled", "true" )
+                );
+    }
+
+    @AfterClass
+    public static void bringDownCluster() {
+        if (cluster != null) {
+            cluster.close();
+        }
+    }
+    
+    // TODO: fabric tests
+
+    @Test
+    public void testTriggerAddAllowedOnlyInSysLeaderMember() {
+        final String query = "CALL apoc.trigger.add($name, 'RETURN 1', {})";
+        triggerInSysLeaderMemberCommon(query, DB_NON_WRITER_ERROR, DEFAULT_DATABASE_NAME);
+    }
+
+    @Test
+    public void testTriggerRemoveAllowedOnlyInSysLeaderMember() {
+        final String query = "CALL apoc.trigger.remove($name)";
+        triggerInSysLeaderMemberCommon(query, DB_NON_WRITER_ERROR, DEFAULT_DATABASE_NAME);
+    }
+
+    @Test
+    public void testTriggerInstallAllowedOnlyInSysLeaderMember() {
+        final String query = "CALL apoc.trigger.install('neo4j', $name, 'RETURN 1', {})";
+        triggerInSysLeaderMemberCommon(query, TRIGGER_NOT_ROUTED_ERROR, SYSTEM_DATABASE_NAME);
+    }
+
+    @Test
+    public void testTriggerDropAllowedOnlyInSysLeaderMember() {
+        final String query = "CALL apoc.trigger.drop('neo4j', $name)";
+        triggerInSysLeaderMemberCommon(query, TRIGGER_NOT_ROUTED_ERROR, SYSTEM_DATABASE_NAME);
+    }
+
+    private static void triggerInSysLeaderMemberCommon(String query, String triggerNotRoutedError, String dbName) {
+        final List<Neo4jContainerExtension> members = cluster.getClusterMembers();
+        assertEquals(4, members.size());
+        for (Neo4jContainerExtension container: members) {
+            // we skip READ_REPLICA members
+            final Driver driver = getDriverIfNotReplica(container);
+            if (driver == null) {
+                continue;
+            }
+            System.out.println("TriggerClusterRoutingTest.triggerInSysLeaderMemberCommon");
+            Session session = driver.session(SessionConfig.forDatabase(dbName));
+            final String address = container.getEnvMap().get("NEO4J_dbms_connector_bolt_advertised__address");
+            if (dbIsWriter(session, dbName, address)) {
+                final String name = UUID.randomUUID().toString();
+                testCall( session, query,
+                        Map.of("name", name),
+                        row -> assertEquals(name, row.get("name")) );
+            } else {
+                try {
+                    testCall(session, query,
+                            Map.of("name", UUID.randomUUID().toString()),
+                            row -> fail("Should fail because of non writer trigger addition"));
+                } catch (Exception e) {
+                    String errorMsg = e.getMessage();
+                    assertTrue("The actual message is: " + errorMsg, errorMsg.contains(triggerNotRoutedError));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testTriggersAllowedOnlyWithAdmin() {
```

Comment:
Could this test be moved to TriggerTest.java and TriggerNewProceduresTest.java? Or else would you mind explaining why it needs to be run inside a cluster, as I'm not sure I know the reason?

Reply:
Actually we don't need a cluster.
I placed this test here because this way we can do `GraphDatabase.driver(uri, AuthTokens.basic(nonAdminUser, nonAdminPwd)))`.

I don't think there is an easy way to do something like that in `TriggerNewProceduresTest` (but I might be wrong) which use a `TestDatabaseManagementServiceBuilder`.

We could also create another single-instance docker test (as mentioned above).

Reply:
Moved to `TriggerEnterpriseFeaturesTest`